dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java

```diff
@@ -318,46 +321,54 @@ private static void levelOrder(
               DependencyNode failedDependencyNode = artifactResult.getRequest().getDependencyNode();
               ExceptionAndPath failure =
                   ExceptionAndPath.create(parentNodes, failedDependencyNode, resolutionException);
-              if (isUnacceptableResolutionException(failure)) {
-                resolutionFailures.add(failure);
+              if (!isSafeResolutionException(failure)) {
+                unsafeResolutionFailures.add(failure);
               }
             }
           } catch (DependencyCollectionException collectionException) {
             DependencyNode failedDependencyNode = collectionException.getResult().getRoot();
             ExceptionAndPath failure =
                 ExceptionAndPath.create(parentNodes, failedDependencyNode, collectionException);
-            if (isUnacceptableResolutionException(failure)) {
-              resolutionFailures.add(failure);
+            if (!isSafeResolutionException(failure)) {
+              unsafeResolutionFailures.add(failure);
             }
           }
         }
       }
+
       for (DependencyNode child : dependencyNode.getChildren()) {
+        // A graph including provided scope may become very big (more than 20,000 nodes).
+        // https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/367
+        // Optimizing the traversal by not visiting a node twice
+        if (includeProvidedScope && !visitedArtifacts.add(child.getArtifact())) {
+          // The optimization is not applicable for graphs used for upper-bound suggestions.
+          continue;
+        }
         @SuppressWarnings("unchecked")
         Stack<DependencyNode> clone = (Stack<DependencyNode>) parentNodes.clone();
         queue.add(new LevelOrderQueueItem(child, clone));
       }
     }
 
-    if (!resolutionFailures.isEmpty()) {
-      throw new AggregatedRepositoryException(resolutionFailures);
+    if (!unsafeResolutionFailures.isEmpty()) {
+      throw new AggregatedRepositoryException(unsafeResolutionFailures);
     }
   }
 
   /**
-   * Returns true if {@code exceptionAndPath.getPath} does not contain {@code optional} dependency
-   * and the path does not contain {@code scope:provided} dependency.
+   * Returns true if {@code exceptionAndPath.getPath} contains an {@code optional} dependency or a
+   * {@code scope:provided} dependency.
    */
-  private static boolean isUnacceptableResolutionException(ExceptionAndPath exceptionAndPath) {
+  private static boolean isSafeResolutionException(ExceptionAndPath exceptionAndPath) {
```

Comment:
I wonder if there's an accepted name for this concept?

Reply:
Perhaps just `isOptionalOrProvided`

Reply:
Renamed.